### DISCUSSION TITLE
How to avoid message loss while sending via a consistent hash based router with local actors?

### DISCUSSION THREAD

**Discussion by sureshchahal** [akkadotnet/akka.net#6944] [Q&A]
I am using Akka.NET to create a consistent hash based router . The router is created locally and use local actors. There is no remote actor, all actors are on a single machine and system. However, I noticed that I am losing 50% of the messages sent to the local actor. This is surprising to me, because I read in [this blog post from petabridge](https://petabridge.com/blog/akkadotnet-at-least-once-message-delivery/) that “All in-memory message passing within one application, for instance, is going to be guaranteed unless you hit an OutOfMemoryException or any of the other standard CLR failures”. I do not see any exception in my application.

How can I ensure that local messages are always delivered exactly once on my local actor?

Here is code snippet that reproduce this issue
` var consistentHashRouter = mySystem.ActorOf(
     Props.Create<ReadAfterWriteLatencyTestActor>(randomNodeRouter, metricCollectorActor)
     .WithRouter(new ConsistentHashingPool(workerCount)),
     "ConcurrentPool");

 

// generate 100 random campaign ids
 Random random = new Random();
 
 for (int j = 0; j < numberOfTimesToRun; j++)
 {
     var testMessage = new PerformReadAfterWriteLatencyTest(random.Next(1000, 10000), 1, 10);
     consistentHashRouter.Tell(testMessage);
     Console.WriteLine($"Sent message for test run {j}: {testMessage.ToString()}");

     if (j % 10000 == 0)
     {
         Console.WriteLine($"Waiting for 5 seconds for akka to catch up");
         System.Threading.Thread.Sleep(5000);
     }
 }`

**Comment by Aaronontheweb** [comment]
hi @sureshchahal 

In Akka.NET, in order to use a consistent hashing router you must either:

* Have all messages sent through that router implement the [IConsistentHashable](https://getakka.net/api/Akka.Routing.IConsistentHashable.html);
* Wrap all messages sent through the router inside a [ConsistentHashableEnvelope](https://getakka.net/api/Akka.Routing.ConsistentHashableEnvelope.html); or
* Declare your consistent hash router programmatically and manually pass in a hashing function which covers all possible inputs. [Here’s an example](https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.Tests/Routing/ConsistentHashingRouterSpec.cs#L142).

Which of these are you using?

**Reply by sureshchahal** [reply]
I have my all messaged implement [IConsistentHashable](https://getakka.net/api/Akka.Routing.IConsistentHashable.html);

This is how my messages are implemented

`internal class PerformReadAfterWriteLatencyTest: TimeSpendInQueueMessageBase, IConsistentHashable
{
    public readonly int CampaignId;
    public readonly int PublisherId;
    public readonly int SpendAmount;
    public readonly int UnixTimeStampSinceEpoch;
    public readonly int SharedBudgetId;

    public object ConsistentHashKey { get { return CampaignId; } }

    internal PerformReadAfterWriteLatencyTest(int campaignId, int publisherId, int spendAmount, int unixTimeStampSinceEpoch)
    {
        CampaignId = campaignId;
        PublisherId = publisherId;
        SpendAmount = spendAmount;
        UnixTimeStampSinceEpoch = unixTimeStampSinceEpoch;
        SharedBudgetId = 0;
    }

    internal PerformReadAfterWriteLatencyTest(int campaignId, int publisherId, int spendAmount)
    {
        CampaignId = campaignId;
        PublisherId = publisherId;
        SpendAmount = spendAmount;
        UnixTimeStampSinceEpoch = (int)DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        SharedBudgetId = 0;
    }
}`

**Reply by Aaronontheweb** [reply]
Ok, let me take a look and see if I can reproduce the issue. Worth asking: is it possible that your actors `randomNodeRouter, metricCollectorActor` are where problems are occurring? Because if the CHR wasn't working this would have been a "hair on fire" bug nearly 10 years ago - let alone today. We'll still take a look, but it'd be worth doing some due diligence on your code too.

**Reply by Aaronontheweb** [reply]
Just tried to reproduce it here:

```csharp
async Task Main()
{
	var myActorSystem = ActorSystem.Create("Repro");
	var poolSize = 25;
	var chr = myActorSystem.ActorOf(Props.Create(() => new ChrTestActor()).WithRouter(new ConsistentHashingPool(poolSize)), "ChrRouters");
	var chrCounter = myActorSystem.ActorOf(Props.Create(() => new ChrTestResultsGatherer(chr, poolSize)), "testResults");
	var testMessage = new PerformReadAfterWriteLatencyTest(Random.Shared.Next(1000, 10000), 1, 10, 1, 1);
	
	for(int i = 0; i < 10_000; i++){
		chr.Tell(testMessage with { CampaignId = Random.Shared.Next(1000, 10000) });
	}
	
	chrCounter.Tell(Finished.Instance);
	
	await myActorSystem.WhenTerminated;
}

internal sealed record PerformReadAfterWriteLatencyTest(int CampaignId, int PublisherId, int SpendAmount, int UnixTimeStampSinceEpoch, int SharedBudgetId) : IConsistentHashable
{
	public object ConsistentHashKey => CampaignId;
}

internal sealed class Finished{
	public static readonly Finished Instance = new();
	private Finished(){}
}

internal sealed class ChrTestResultsGatherer : ReceiveActor{
	private readonly int _actorCount;
	private readonly IActorRef _chrRouter;
	private readonly ILoggingAdapter _log = Context.GetLogger();
	
	public ChrTestResultsGatherer(IActorRef chrRouter, int actorCount){
		_actorCount = actorCount;
		_chrRouter = chrRouter;
		
		Waiting();
	}
	
	public void Waiting(){
		Receive<Finished>(f => {
			// ping all recipients
			_chrRouter.Tell(new Broadcast(Finished.Instance));
			
			Become(() => Counting(_actorCount));
		});
	}
	
	public void Counting(int remainingActors){
		var dict = new Dictionary<IActorRef, int>();
		ReceiveAsync<int>(async i => {
			dict[Sender] = i;
			if (dict.Count == _actorCount)
			{
				var c = 0;
				foreach (var d in dict) {
					_log.Info("Actor [{0}]: {1}", c, d.Value);
					c++;
				}
				
				var totalMsgsProcessed = dict.Values.Sum();
				_log.Info("Total messages processed: [{0}]", totalMsgsProcessed);
				
				await Context.System.Terminate();
			}
		});
	}
}

internal sealed class ChrTestActor : ReceiveActor{
	private int _count = 0;
	private readonly ILoggingAdapter _log = Context.GetLogger();
	
	public ChrTestActor(){
		Receive<PerformReadAfterWriteLatencyTest>(r => {
			_count++;	
		});
		
		Receive<Finished>(_ => {
			Sender.Tell(_count);
		});
	}
}
```

Produced the following:

```
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [0]: 448
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [1]: 323
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [2]: 302
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [3]: 340
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [4]: 163
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [5]: 448
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [6]: 418
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [7]: 493
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [8]: 220
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [9]: 449
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [10]: 402
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [11]: 491
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [12]: 457
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [13]: 433
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [14]: 330
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [15]: 511
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [16]: 275
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [17]: 463
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [18]: 478
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [19]: 598
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [20]: 238
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [21]: 515
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [22]: 459
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [23]: 393
[INFO][10/04/2023 19:26:56.523Z][Thread 0060][akka://Repro/user/testResults] Actor [24]: 353
[INFO][10/04/2023 19:26:56.524Z][Thread 0060][akka://Repro/user/testResults] Total messages processed: [10000]
```

**Reply by Aaronontheweb** [reply]
Like I said - very unlikely this is an Akka.NET issue; that code has been used very heavily for many years.

**Reply by sureshchahal** [reply]
If we assume that due to some issue, one of worker of CHR is restarted, wouldn't CHR create a new worker and process remaining messages? or we just lose all message pending in queue of worker that stopped?

Also I reading between lines, it seems that it is guarantee that no message is lost when sent to local actor either directly or via router? Is this right assumption? It is quite important for me as I am trying to build a distributed application that keep track of Budget spend in real time.

**Reply by Aaronontheweb** [reply]
Which version of Akka.NET are you using? Because there was a router / routee supervision bug we fixed late last year in Akka.NET v1.4.50: https://github.com/akkadotnet/akka.net/issues/6295 - you could be running into that issue if you are using an older version of Akka.NET.

**Reply by Aaronontheweb** [reply]
> Also I reading between lines, it seems that it is guarantee that no message is lost when sent to local actor either directly or via router? Is this right assumption? It is quite important for me as I am trying to build a distributed application that keep track of Budget spend in real time.

Messages can only be lost, when working with local actors, if the actor is terminated before it has a chance to process the message (via `Context.Stop` or a `Directive.Stop` or by a parent actor restarting, which was the router issue we fixed in v1.4.50.) All in-memory messaging is facilitated by `ConcurrentQueue<T>` under the covers.

**Comment by sureshchahal** [comment]
I am using 1.5.12.

So does it mean that all enqueued messages are lost too on restart? I
though restart use same message queue, so messages are not actually lost
(except atmost 1 in flight message). This is interesting if restart means
to have a new queue and hence all messages are lost

**Reply by Aaronontheweb** [reply]
> So does it mean that all enqueued messages are lost too on restart?

No - messages are not lost on restart: https://www.youtube.com/watch?v=CC2XdYlpmvo

**Reply by Aaronontheweb** [reply]
Restart != Terminated

Restart == transient error handling + recovery
Terminated == permanent deletion actor's resources + mailbox and invalidation of the `IActorRef`

**Reply by Aaronontheweb** [reply]
When a parent actor restarts, by default of its children will be terminated - thus, having parent actors carry out work that can result in exceptions being thrown is considered to be code smell in Akka.NET usually.

**Comment by sureshchahal** [comment]
Thank you for the clarification and appreciate the quick responses . Let me
add more logging to find if something is causing termination and thus
message loss.

**Reply by Aaronontheweb** [reply]
> > This is interesting if restart means to have a new queue and hence all messages are lost


Watch the YouTube video - we explicitly show in there that restarting doesn't affect messages that are in the queue.
